pkg/ciliumidentity: Add Cilium Identity Controller Operator hive cell

ovidiutirla opened the pull request:
Add the reconciler logic for reconciling CIDs, Pods, Namespaces.

The Cilium Identity API objects will be managed by the cilium-operator instead of the cilium-agent. The desired state for Cilium Identities is determined by the cilium-operator based on monitored events for Cilium Identities, Pods, Namespaces, and Cilium Endpoint Slices (if enabled). When pod updates occur for a unique label set derived from pod and namespace labels, the cilium-operator creates Cilium Identities.

Related CFP #27752
Draft full implementation: https://github.com/cilium/cilium/pull/33204

ovidiutirla committed:
pkg/logging: Add CIDName to log fields

The field will be used mainly by operator managing CIDs.
Related #27752

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/watchers: Add helper func to check CE store for CID using index

Add helper function HasCEWithIdentity to check if CiliumEndpoint store contains an endpoint with a given identity.

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/ciliumidentity: Add CID Controller Operator cell and reconciler

Add the CID cell for operator with a reconciler logic for reconciling CIDs, Pods, Namespaces.

Signed-off-by: Ovidiu Tirla <otirla@google.com>

dlapcevic reviewed (approved):
lgtm

pippolo84 commented on operator/identitygc/crd_gc.go:
```diff
@@ -8,11 +8,12 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/cilium/cilium/operator/k8s"
```
Nit: we enforce the following import order

- Go stdlib packages
- external packages
- cilium/cilium packages

So this one should be in the last set (starting from line 17).
Actually, one of the linters should have complained about this, I'm not sure why it didn't.

ovidiutirla replied:
This was formatted with `goimports` but adjusted now manually.

pippolo84 commented on operator/pkg/ciliumidentity/cache.go:
```diff
@@ -162,31 +168,20 @@ func (c *CIDUsageInPods) AssignCIDToPod(podName, cidName string) (string, int) {
-	return cidName, count, nil
-}
-
-func (c *CIDUsageInPods) CIDUsedByPod(podName string) (string, bool) {
```
I think it would be better to break this commit into three separate commits:

- one to remove `CIDUsedByPod`
- one to add the logger to NewCIDState
- one to change `RemovePod` return type

ovidiutirla replied:
Done.

pippolo84 commented on operator/k8s/resources.go:
```diff
@@ -66,3 +66,16 @@ type Resources struct {
+// HasCEWithIdentity returns true or false if the Cilium Endpoint store has
+// the given identity.
+// Once the CiliumEndpoint is migrated to use resources
+// the cepStore can be removed and used the local store.
```
I'm not sure I'm getting this part right: CiliumEndpoint in the operator is already using `Resource[T]` framework.

ovidiutirla replied:
Indeed, you're right! Removed the comment.

pippolo84 commented on operator/pkg/ciliumidentity/cache_test.go:
```diff
@@ -5,15 +5,17 @@ package ciliumidentity
 	"github.com/stretchr/testify/assert"
 
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging"
```
Nit: move this imports with the the other ones below

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/cilium/cilium/pkg/labels"
```
Ditto

ovidiutirla replied:
done

pippolo84 replied:
I still see `pkg/option` and `pkg/labels` imports at the same rows, maybe you forgot to push the changes?

ovidiutirla replied:
should be done now, not sure exactly how it got reformatted back to the old format.

pippolo84 commented on operator/pkg/ciliumidentity/cell.go:
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
```
I don't think there's much value in breaking the controller logic into this commit (https://github.com/cilium/cilium/pull/33383/commits/82af79bdf79ce396ee5751676189ebb8f65716f7) and the next one (https://github.com/cilium/cilium/pull/33383/commits/7079ecb4fb6bca2fb683381f813c2b4030b8e6c9). Could you please merge them?

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+	podStore, _ := pod.Store(ctx)
+	cidStore, _ := ciliumIdentity.Store(ctx)
+	cepStore, _ := ciliumEndpoint.Store(ctx)
+	cesStore, _ := ciliumEndpointSlice.Store(ctx)
```
Is there a reason to swallow all the errors?

ovidiutirla replied:
checked for all the errors and returned in case of failure to initialize

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -0,0 +1,77 @@
+
+func (c *Controller) Stop(hookContext cell.HookContext) error {
+	//TODO implement me
+	panic("implement me")
```
I know that this is just a placeholder, but there's no need to use a `panic`, a `return nil` is enough. Also, instead of the comment, I suggest to write something in the commit description to clarify that this controller will be linked later into the operator hive.

ovidiutirla replied:
Done.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+			}
+		}
+		return nil
+	}
```
Since the current implementation of the GC is not the one proposed in PR https://github.com/cilium/cilium/pull/33380, this change should be thought to work with the existing implementation in order for this PR to be merged independently.

In case the pod is not found, we should simply early-returning to avoid allocating a CID.

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+		},
+	}
+
+	ns, exists, err := r.nsStore.Get(nsLookupObj)
```
Shouldn't we use `r.nsStore.GetByKey(resource.Key{Name: namespace})` instead?

ovidiutirla replied:
updated.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+	prevCIDName, count := r.cidUsageInPods.AssignCIDToPod(podName, cidName)
+
+	r.logger.Info("CID allocated for pod",
+		logfields.K8sPodName, fmt.Sprintf("%s/%s", pod.Namespace, pod.Name),
```
I expect the CID allocation to happen pretty often, so it might be better to remove `fmt.Sprintf` in favor of a simple `+` concat (see https://github.com/cilium/cilium/issues/19571)

ovidiutirla replied:
I propose we maintain the current implementation. Utilizing `+` for string concatenation would necessitate creating additional strings due to Go's immutable strings. The linked issue suggests employing string builders or buffers, but incorporating them here might introduce unnecessary complexity and reduce code readability. We can always adjust log verbosity later if it becomes a performance concern.

pippolo84 replied:
> Utilizing + for string concatenation would necessitate creating additional strings due to Go's immutable strings.

That's correct, but very often the concatenation is still faster.

> We can always adjust log verbosity later if it becomes a performance concern.

Yep, I agree. No need to micro-optimize ahead of time, let's resolve this one :+1:

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+
+	if len(prevCIDName) > 0 && count == 0 && !r.cidIsUsedInCEPOrCES(prevCIDName) {
+		_ = r.handleCIDDeletion(prevCIDName)
+	}
```
Same here, right now the CID deletion is done through the "old" identity gc, so this is not needed.

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+		_ = r.handleCIDDeletion(prevCIDName)
+	}
+
+	if isNewCID {
+		r.queueOps.enqueueCIDReconciliation(cidResourceKey(cidName), 0)
+	}
```
Since `enqueueCIDReconciliation` is not implemented, let's add this once there will be actual logic to be run.

ovidiutirla replied:
I think having the `enqueueCIDReconciliation` is an important part of this PR and should be kept to provide context on how we plan to make the operator allocate CIDs for pods. We can currently think about this as an abstract queue that will be later implemented.

pippolo84 replied:
I took a look at the draft implementation in https://github.com/cilium/cilium/pull/33204
What about removing the interface and pass the reconciler a couple of callbacks, once for CID reconciliation and the other one for Pod reconciliation?
I am not seeing other implementations of the interface and the callbacks should allow equal flexibility (if not greater) in testing. WDYT?

ovidiutirla replied:
IMO interfaces provide stronger contracts, type safety, and facilitate easier refactoring compared to callbacks. Even if we don't anticipate multiple reconciler implementations now, using an interface allows extensibility and maintainability for the future. Additionally, using callbacks could make the solution less readable.

pippolo84 replied:
Let's leave it as is for now, we can still re-evaluate this later when the full implementation will be in.

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -66,7 +78,40 @@ func registerController(p params) {
+	panic("implement me")
+}
+
+func (c *Controller) enqueuePodReconciliation(podKey resource.Key, delay time.Duration) {
```
I suggest to avoid adding those placeholders and simply add this logic in the next PRs where it is actually needed.

ovidiutirla replied:
These are required for the `newReconciler` as it expects to find the implementation of the controller. Removed the TODOs and panic though.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler_test.go:
```diff
@@ -0,0 +1,745 @@
+		//	initialStoreState:   testCreateCIDObj(cidName, testLbsA),
+		//	errorDuringDeletion: true,
+		//	expectedQueueSize:   1,
+		//},
```
Same here, let's remove this for now, we'll think about re-adding these cases with the identity GC refactor PR.

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler_test.go:
```diff
@@ -0,0 +1,745 @@
+		//		queueSize:     1,
+		//		existsInQueue: []string{cidName},
+		//	},
+		//},
```
Ditto

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+// not the same.
+func (r *reconciler) reconcileCID(cidResourceKey resource.Key) error {
+	cidName := cidResourceKey.Name
+	storeCIDObj, existsInStore, err := r.cidStore.CacheStore().GetByKey(cidResourceKey.Name)
```
Why calling `CacheStore()` to get the underlying store?

ovidiutirla replied:
updated to use cacheStore directly.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+	}
+
+	var storeCID *cilium_api_v2.CiliumIdentity
+	if existsInStore {
```
I find this function pretty hard to read. We should try to break into two cases (like CID in store/CID not in store or alternatively CID in desired state/CID not in desired state) and call a separate helper to handle each one.

ovidiutirla replied:
No longer required if we are not using CacheStore()

pippolo84 replied:
I think that the function has not improved much removing the usage of the `CacheStore()`. Moreover, calling `handleCIDDeletion` is pointless, as in the current implementation, there is another cell taking care of the identities GC. Finally, the `handleCIDDeletion` placeholder should be removed and, in case, it will be added later in PR https://github.com/cilium/cilium/pull/33380 (and possible follow-ups).

ovidiutirla replied:
removed the `handleCIDDeletion`

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+func (r *reconciler) createCID(cidName string, cidKey *key.GlobalIdentity) error {
+	cidLabels := cidKey.GetAsMap()
+	selectedLabels, skippedLabels := identitybackend.SanitizeK8sLabels(cidLabels)
+	r.logger.Debug("Skipped non-kubernetes labels when labelling CID. All labels will still be used in identity determination", logfields.Labels, skippedLabels)
```
Shouldn't we guard this log checking the length of `skippedLables` first? Same for the call in `updateCID`.

ovidiutirla replied:
The current logger is configured for Debug only, but I believe we can utilize all labels for debugging purposes without requiring extra logic to verify their length.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,474 @@
+
+	r.logger.Info("Updating CID", logfields.CIDName, cid.Name)
+
+	_, err := r.clientset.CiliumV2().CiliumIdentities().Update(context.TODO(), cid, metav1.UpdateOptions{})
```
We should use the reconciler context here, same for the other usages of the clientset.

ovidiutirla replied:
done

pippolo84 reviewed (changes requested):
Thanks for the PR!

Did a first pass and left some comments. Let's address these before going ahead.

ovidiutirla committed:
operator/ciliumidentity: Cleanup unused CIDUsedByPod method

CIDUsedByPod is not planned to be used in operator managing CIDs

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
operator/ciliumidentity: Update CIDState to use slog

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
operator/ciliumidentity: Change RemovePod method signature

Modified the error parameter to a boolean type since the sole error condition was whether the pod existed in the cache or not.

Signed-off-by: Ovidiu Tirla <otirla@google.com>

pippolo84 commented on operator/pkg/ciliumidentity/reconciler_test.go:
```diff
@@ -0,0 +1,721 @@
+				t.Errorf("Unexpected fakeWorkQueue size: got %v, want %v", len(queueOps.fakeWorkQueue), tc.expectedQueueSize)
+			}
+
+			// TODO check for CID deletion too, adapt once we decide how to handle CID deletion
```
Could you please add a reference to the Identity GC PR where the discussion is taking place? That will give useful context to a future casual reader.

ovidiutirla replied:
Added a link to the other PR.

pippolo84 reviewed (commented):
Thanks for the follow up. :+1: 

Quickly reviewed the latest changes and left a couple of additional comments. I'll do another pass later.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+	cidUsageInPods  *CIDUsageInPods
+	cidUsageInCES   *CIDUsageInCES
+	// Ensures no CID duplicates are created while allocating CIDs in parallel,
+	// and avoids race conditions when CIDs are being deleted.
```
```suggestion
	// Ensures no CID duplicates are created while allocating CIDs in parallel
	// (that is, when processing a pod event and a CID event concurrently)
	// and avoids race conditions when CIDs are being deleted.
```

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+
+// syncPodsOnStartup ensures that all pods have a CID for their labels, and that
+// all non-used CIDs are deleted. Non used CIDs are those that aren't in use by
+// any of the pods and also don't exist in CESs (if CES is enabled).
```
Shouldn't the part regarding deleting the CIDs be removed given that we'll stick to the old GC implementation, for now?

ovidiutirla replied:
updated the docs

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+
+	cidKey, existsInDesiredState := r.desiredCIDState.LookupByID(cidName)
+	if !existsInDesiredState && !existsInStore {
+		_ = r.makeIDAvailable(cidName)
```
Here we should be able to differentiate between the cases:

- `makeIDAvailable` failed because the identity was not previously reserved in the allocator. Nothing to do in that case.
- `makeIDAvailable` failed for unexpected reasons. Here we should propagate the error

ovidiutirla replied:
I've added a log message to track this, I don't believe we need to propagate the error from reconcileCID.
If it would throw an error, we'll simply re-queue the cid reconciliation and it'll fail again. 

The potential failure reasons:
1) cidName is not an int (shouldn't happen)
2) cid is not in the range (shouldn't happen)
3) cid is there already so we shouldn't anyway propagate it (shouldn't happen)

pippolo84 replied:
Let's go with the warning, then! :+1:

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+// the following:
+// 1. Nothing - If CID doesn't exist in both desired state cache and watcher's
+// store.
+// 2. Deletes CID - If CID only exists in the watcher's store, and it isn't used.
```
As above, this comment needs to be updated given that we are not deleting CIDs here.

ovidiutirla replied:
updated the docs

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+	cidIsUsed := r.cidIsUsedInPods(cidName) || r.cidIsUsedInCEPOrCES(cidName)
+	if !existsInDesiredState {
+		return nil
+	}
```
Couple of questions:

1) Why are we checking this here after calculating `cidIsUsed`? 

2) My understanding Is that the case `existsInDesiredState == false && cidIsUsed == true` is possible (or should be possible) only when the agent is managing CIDs together with the operator (cfr [here](https://docs.google.com/document/d/1Hcc_2mB9OOUxrqQgZ-gSYDPnLYE_If_TCzVbUGDOdGM/edit#heading=h.f1w8m5hi2mkd) in the CfP). Is my assumption correct? If so, shouldn't we mark the identity as "not available" in the basic ID allocator, to avoid the operator allocating that ID?

ovidiutirla replied:
1. refactored the if statements to improve redability
2. Added your suggestion. This scenario is unlikely to occur in real situations because the order of events should prevent it. The operator typically reacts to the creation of a pod object before the agent has a chance to process the corresponding pod event.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+
+	if !cidIsUsed {
+		if existsInStore {
+			return nil
```
Might be worth adding a comment here stating that the identity GC will eventually delete the CID and that there's nothing more to do on the reconciler side, for now.

ovidiutirla replied:
added to the methods docs.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+func (r *reconciler) handleStoreCIDMatch(storeCIDs []*cilium_api_v2.CiliumIdentity) (string, error) {
+	if len(storeCIDs) == 0 {
+		return "", fmt.Errorf("store CIDs list is empty")
+	}
```
We already checked this before calling the function. Since `handleStoreCIDMatch` is called just once, there's no value in checking again `storeCIDs`.

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,444 @@
+		if toDelete {
+			r.queueOps.enqueueCIDReconciliation(cidResourceKey(cid.Name), 0)
+		}
+	}
```
We are not deleting identities from the new controller, so is the call to `enqueueCIDReconciliation` (and the `toDelete` logic) necessary?

ovidiutirla replied:
refactored the method as we no longer need to reconcile all CIDs if operator managing CIDs is no longer deleting the CIDs.

pippolo84 reviewed (changes requested):
Thanks for the follow up.

Main concern is around some leftover logic that I think is not needed until we're gonna refactor the identity GC. Please be sure to rework every part that might assume a different GC implementation before requesting another review. Thanks! :pray:

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -0,0 +1,438 @@
+
+		}
+		return nil
+	}
```
Nonblocking: using an early return here makes the code a bit nicer

```suggestion
	if !existsInDesiredState {
		if !cidIsUsed {
			return nil
		}

		r.cidCreateLock.Lock()
		defer r.cidCreateLock.Unlock()

		id, err := r.idAllocator.ValidateIDString(cidName)
		if err != nil {
			return err
		}
		return r.idAllocator.Allocate(idpool.ID(id))
	}
```

ovidiutirla replied:
refactored

pippolo84 commented on operator/pkg/ciliumidentity/reconciler_test.go:
```diff
@@ -0,0 +1,715 @@
+				t.Errorf("Expected no CIDs to be used when CES is disabled, but found %d", len(reconciler.cidUsageInCES.cidUsageCount))
+			}
+		})
+	}
```
Since we are using table driven sub-tests and we just need to compare two maps, we can avoid differentiating the test logic with an if an do this instead:

```go
t.Run(tc.name, func(t *testing.T) {
	...

	reconciler.syncCESsOnStartup()

	if !reflect.DeepEqual(reconciler.cidUsageInCES.cidUsageCount, tc.expectedCIDs) {
		t.Errorf("Expected CID usage count map to be %v, but got %v", tc.expectedCIDs, reconciler.cidUsageInCES.cidUsageCount)
	}
})
```

(for this to work you also have to change the `expectedCIDs` in the first subtest to be `map[int64]int{}`, otherwise you can use something like https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#EquateEmpty)

ovidiutirla replied:
done

pippolo84 commented on operator/pkg/ciliumidentity/reconciler_test.go:
```diff
@@ -0,0 +1,715 @@
+		name          string
+		cidList       []*capi_v2.CiliumIdentity
+		expectedCID   string
+		expectedError bool
```
There are no sub-tests where we expect an error, so we can remove this and simply check for no error at every run.

ovidiutirla replied:
removed expectedError

pippolo84 reviewed (changes requested):
This looks better, thanks! :100: 

Just a final couple of comments on the tests (nothing major) and a non-blocking suggestion to make the reconciling main function a bit nicer.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler_test.go:
```diff
@@ -0,0 +1,716 @@
+			//	}
+			//} else if len(reconciler.cidUsageInCES.cidUsageCount) != 0 {
+			//	t.Errorf("Expected no CIDs to be used when CES is disabled, but found %d", len(reconciler.cidUsageInCES.cidUsageCount))
+			//}
```
Leftover commented code

ovidiutirla replied:
removed

pippolo84 reviewed (approved):
LGTM, thanks! :rocket:

pchaigno reviewed (approved):
Approving for the 2 lines of "CLI" change (logfield).
